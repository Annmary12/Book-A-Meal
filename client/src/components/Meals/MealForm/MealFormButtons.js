import React from 'react';
import PropTypes from 'prop-types';
import styles from './meal-form.scss';

const ModalButtons = ({ isSubmitting, handleSubmit, handleClose }) => (
  <div className={styles.btns}>
    <button
      type="button"
      disabled={isSubmitting ? 'disabled' : ''}
      className={styles.successBtn}
      onClick={handleSubmit}
    >
      {isSubmitting ? 'Please wait..' : 'Save'}
    </button>
    <button
      type="button"
      className={styles.closeModalBtn}
      onClick={handleClose}
    >
      Cancel
    </button>

  </div>
);

ModalButtons.propTypes = {
  isSubmitting: PropTypes.bool.isRequired,
  handleSubmit: PropTypes.func.isRequired,
  handleClose: PropTypes.func.isRequired
};

export default ModalButtons;
