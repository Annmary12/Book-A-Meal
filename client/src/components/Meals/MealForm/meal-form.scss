@import '../../../../assets/scss/_variables';
@import '../../../../assets/scss/form';
@import '../../../../assets/scss/buttons';