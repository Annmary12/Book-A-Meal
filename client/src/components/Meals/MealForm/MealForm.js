import React, { Component } from 'react';
import PropTypes from 'prop-types';
import InputField from '../../Form/InputField.js';
import MealFormButtons from './MealFormButtons.js';
import styles from './meal-form.scss';

/**
 * @class MealForm
 */
class MealForm extends Component {
  static propTypes = {
    values: PropTypes.objectOf(PropTypes.string),
    touched: PropTypes.objectOf(PropTypes.bool),
    errors: PropTypes.objectOf(PropTypes.string),
    isSubmitting: PropTypes.bool,
    handleChange: PropTypes.func.isRequired,
    handleBlur: PropTypes.func,
    handleSubmit: PropTypes.func.isRequired,
    handleClose: PropTypes.func.isRequired,
    setFieldValue: PropTypes.func.isRequired
  };

  static defaultProps = {
    values: {},
    touched: {},
    errors: {},
    isSubmitting: false,
    handleBlur() {}
  };

  handleImageChange = (event) => {
    this.props.setFieldValue('image', event.currentTarget.files[0]);
  }

  /**
   * @returns {JSX} React JSX
   */
  render() {
    return (
      <form onSubmit={this.props.handleSubmit}>
        <h2>Meal details</h2>
        <InputField name="title" label="Title" {...this.props} />
        <InputField name="description" label="Description" {...this.props} />
        <InputField name="price" label="Price" type="number" {...this.props} />
        <label htmlFor="image" className={styles.label}>
          {this.props.values.img ? 'Change image:' : 'Upload image:' }
          <input
            name="image"
            type="file"
            className={styles.fileInput}
            onChange={this.handleImageChange}
            onBlur={this.props.handleBlur}
          />
        </label>
        { this.props.values.img && <img src={this.props.values.img} width="80" alt="meal" /> }
        <MealFormButtons
          isSubmitting={this.props.isSubmitting}
          handleClose={this.props.handleClose}
          handleSubmit={this.props.handleSubmit}
        />
      </form>
    );
  }
}

export default MealForm;
