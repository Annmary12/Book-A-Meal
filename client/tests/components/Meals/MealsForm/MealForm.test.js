import React from 'react';
import MealForm from '../../../../src/components/Meals/MealForm/MealForm';

const props = {
  values: {
    img: undefined
  },
  touched: {},
  errors: {},
  isSubmitting: false,
  handleChange: jest.fn(),
  handleBlur: jest.fn(),
  handleSubmit: jest.fn(),
  handleClose: jest.fn(),
  setFieldValue: jest.fn()
}
describe('MealForm component', () => {
  it('should match snapshot if values.img is undefined', () => {
    const wrapper = shallow(
      <MealForm
        {...props}
      />
    );

    expect(wrapper).toMatchSnapshot();
  });
  it('should match snapshot if values.img is not undefined', () => {
    const wrapper = shallow(
      <MealForm
        {...props}
        values={{ img: 'pic.jpg' }}
      />
    );

    expect(wrapper).toMatchSnapshot();
  });

  describe('handleImageChange method', () => {
    const event = {
      currentTarget: {
        files: ['test-file']
      }
    };

    it('should call setFieldValue prop function with image file', () => {
      const wrapper = shallow(
        <MealForm
          {...props}
        />
      );

      wrapper.instance().handleImageChange(event);

      expect(props.setFieldValue).toHaveBeenCalledWith(
        'image',
        event.currentTarget.files[0]
      )
    })
  })
});