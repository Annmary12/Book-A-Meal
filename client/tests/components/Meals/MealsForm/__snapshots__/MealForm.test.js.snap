// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MealForm component should match snapshot if values.img is not undefined 1`] = `
<form
  onSubmit={[MockFunction]}
>
  <h2>
    Meal details
  </h2>
  <InputField
    errors={Object {}}
    handleBlur={[MockFunction]}
    handleChange={[MockFunction]}
    handleClose={[MockFunction]}
    handleSubmit={[MockFunction]}
    isSubmitting={false}
    label="Title"
    name="title"
    setFieldValue={[MockFunction]}
    touched={Object {}}
    type="text"
    values={
      Object {
        "img": "pic.jpg",
      }
    }
  />
  <InputField
    errors={Object {}}
    handleBlur={[MockFunction]}
    handleChange={[MockFunction]}
    handleClose={[MockFunction]}
    handleSubmit={[MockFunction]}
    isSubmitting={false}
    label="Description"
    name="description"
    setFieldValue={[MockFunction]}
    touched={Object {}}
    type="text"
    values={
      Object {
        "img": "pic.jpg",
      }
    }
  />
  <InputField
    errors={Object {}}
    handleBlur={[MockFunction]}
    handleChange={[MockFunction]}
    handleClose={[MockFunction]}
    handleSubmit={[MockFunction]}
    isSubmitting={false}
    label="Price"
    name="price"
    setFieldValue={[MockFunction]}
    touched={Object {}}
    type="number"
    values={
      Object {
        "img": "pic.jpg",
      }
    }
  />
  <label
    className="label"
    htmlFor="image"
  >
    Change image:
    <input
      className="fileInput"
      name="image"
      onBlur={[MockFunction]}
      onChange={[Function]}
      type="file"
    />
  </label>
  <img
    alt="meal"
    src="pic.jpg"
    width="80"
  />
  <MealFormButtons
    handleClose={[MockFunction]}
    handleSubmit={[MockFunction]}
    isSubmitting={false}
  />
</form>
`;

exports[`MealForm component should match snapshot if values.img is undefined 1`] = `
<form
  onSubmit={[MockFunction]}
>
  <h2>
    Meal details
  </h2>
  <InputField
    errors={Object {}}
    handleBlur={[MockFunction]}
    handleChange={[MockFunction]}
    handleClose={[MockFunction]}
    handleSubmit={[MockFunction]}
    isSubmitting={false}
    label="Title"
    name="title"
    setFieldValue={[MockFunction]}
    touched={Object {}}
    type="text"
    values={
      Object {
        "img": undefined,
      }
    }
  />
  <InputField
    errors={Object {}}
    handleBlur={[MockFunction]}
    handleChange={[MockFunction]}
    handleClose={[MockFunction]}
    handleSubmit={[MockFunction]}
    isSubmitting={false}
    label="Description"
    name="description"
    setFieldValue={[MockFunction]}
    touched={Object {}}
    type="text"
    values={
      Object {
        "img": undefined,
      }
    }
  />
  <InputField
    errors={Object {}}
    handleBlur={[MockFunction]}
    handleChange={[MockFunction]}
    handleClose={[MockFunction]}
    handleSubmit={[MockFunction]}
    isSubmitting={false}
    label="Price"
    name="price"
    setFieldValue={[MockFunction]}
    touched={Object {}}
    type="number"
    values={
      Object {
        "img": undefined,
      }
    }
  />
  <label
    className="label"
    htmlFor="image"
  >
    Upload image:
    <input
      className="fileInput"
      name="image"
      onBlur={[MockFunction]}
      onChange={[Function]}
      type="file"
    />
  </label>
  <MealFormButtons
    handleClose={[MockFunction]}
    handleSubmit={[MockFunction]}
    isSubmitting={false}
  />
</form>
`;
